k12.c

```diff
@@ -0,0 +1,154 @@
+	st[23] = ROTL64(st[15] ^ bc[0], 41);
+	st[15] = ROTL64(st[ 4] ^ bc[4], 27);
+	st[ 4] = ROTL64(st[24] ^ bc[4], 14);
+	st[24] = ROTL64(st[21] ^ bc[1],  2);
+	st[21] = ROTL64(st[ 8] ^ bc[3], 55);
+	st[ 8] = ROTL64(st[16] ^ bc[1], 45);
+	st[16] = ROTL64(st[ 5] ^ bc[0], 36);
+	st[ 5] = ROTL64(st[ 3] ^ bc[3], 28);
+	st[ 3] = ROTL64(st[18] ^ bc[3], 21);
+	st[18] = ROTL64(st[17] ^ bc[2], 15);
+	st[17] = ROTL64(st[11] ^ bc[1], 10);
+	st[11] = ROTL64(st[ 7] ^ bc[2],  6);
+	st[ 7] = ROTL64(st[10] ^ bc[0],  3);
+	st[10] = tmp;
+
+	// Chi
+	bc[0] = st[ 0]; bc[1] = st[ 1]; st[ 0] ^= (~bc[1]) & st[ 2]; st[ 1] ^= (~st[ 2]) & st[ 3]; st[ 2] ^= (~st[ 3]) & st[ 4]; st[ 3] ^= (~st[ 4]) & bc[0]; st[ 4] ^= (~bc[0]) & bc[1];
+	bc[0] = st[ 5]; bc[1] = st[ 6]; st[ 5] ^= (~bc[1]) & st[ 7]; st[ 6] ^= (~st[ 7]) & st[ 8]; st[ 7] ^= (~st[ 8]) & st[ 9]; st[ 8] ^= (~st[ 9]) & bc[0]; st[ 9] ^= (~bc[0]) & bc[1];
+	bc[0] = st[10]; bc[1] = st[11]; st[10] ^= (~bc[1]) & st[12]; st[11] ^= (~st[12]) & st[13]; st[12] ^= (~st[13]) & st[14]; st[13] ^= (~st[14]) & bc[0]; st[14] ^= (~bc[0]) & bc[1];
+	bc[0] = st[15]; bc[1] = st[16]; st[15] ^= (~bc[1]) & st[17]; st[16] ^= (~st[17]) & st[18]; st[17] ^= (~st[18]) & st[19]; st[18] ^= (~st[19]) & bc[0]; st[19] ^= (~bc[0]) & bc[1];
+	bc[0] = st[20]; bc[1] = st[21]; st[20] ^= (~bc[1]) & st[22]; st[21] ^= (~st[22]) & st[23]; st[22] ^= (~st[23]) & st[24]; st[23] ^= (~st[24]) & bc[0]; st[24] ^= (~bc[0]) & bc[1];
+
+	// Iota
+	st[0] ^= RndConst;
+}
+
+uint64_t k12_PoW(const void *input)
+{
+	uint64_t st[25] __attribute__((aligned(64)));
+	uint64_t tmp0, tmp3, tmp4;
+
+	memset(st, 0x00, 200);
+	memcpy(st, input, 80);
+
+	// Padding
+	((uint8_t *)st)[81] = 0x07;
+	st[20] = 0x8000000000000000ULL;
+
+	for(int i = 0; i < 11; ++i) KeccakF1600(st, KeccakF1600RndConsts[i]);
+
+	tmp0 = st[4] ^ st[9] ^ st[14] ^ st[19] ^ st[24] ^ ROTL64(st[1] ^ st[6] ^ st[11] ^ st[16] ^ st[21], 1);
+	tmp3 = st[2] ^ st[7] ^ st[12] ^ st[17] ^ st[22] ^ ROTL64(st[4] ^ st[9] ^ st[14] ^ st[19] ^ st[24], 1);
+	tmp4 = st[3] ^ st[8] ^ st[13] ^ st[18] ^ st[23] ^ ROTL64(st[0] ^ st[5] ^ st[10] ^ st[15] ^ st[20], 1);
+
+	tmp0 ^= st[0];
+	tmp3 = ROTL64(st[18] ^ tmp3, 21);
+	tmp4 = ROTL64(st[24] ^ tmp4, 14);
+
+	return(tmp3 ^ ((~tmp4) & tmp0));
+}
+
+void k12_hash(void *output, const void *input)
+{
+	uint64_t st[25] __attribute__((aligned(64)));
+
+	memset(st, 0x00, 200);
+	memcpy(st, input, 80);
+
+	// Padding
+	((uint8_t *)st)[81] = 0x07;
+	st[20] = 0x8000000000000000ULL;
+
+	for(int i = 0; i < 12; ++i) KeccakF1600(st, KeccakF1600RndConsts[i]);
+	memcpy(output, st, 32);
+}
+
+int scanhash_k12(int thr_id, uint32_t *restrict pdata, int dlen, const uint64_t *restrict ptarget,
+			 uint64_t max_nonce, uint64_t *restrict hashes_done)
+{
+	uint64_t *nonceptr = (uint64_t*) (((char*)pdata) + 39);
+	uint64_t n = *nonceptr - 1;
+	const uint64_t first_nonce = n + 1;
+	const uint64_t Htarg = ptarget[7];
+
+	do
+	{
+		*nonceptr = ++n;
+		uint64_t qword3 = k12_PoW(pdata);
```

Comment:
nice work here dropping the 4th qword, just need to add "dlen" parameter to k12_PoW().